File: apps/mobile/ios/Sources/FlagshipUI/Screens/CompanionDockScreen.swift
import SwiftUI
import FlagshipAPI
import FlagshipCore

/// Phone side of the desktop-initiated remote ceremony.
public struct CompanionDockScreen: View {
    @Environment(\.colorScheme) private var scheme
    @Bindable var vm: CompanionDockViewModel
    let initialApprovalLink: String?

    @State private var pendingRevoke: CompanionSummary?
    @State private var nowMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
    @State private var pastedLink = ""
    @State private var showScanner = false
    @State private var scanError: String?

    public init(vm: CompanionDockViewModel, initialApprovalLink: String? = nil) {
        self.vm = vm
        self.initialApprovalLink = initialApprovalLink
    }

    public var body: some View {
        let c = FSColors.scheme(scheme)
        ScrollView {
            VStack(alignment: .leading, spacing: FS.space.s6) {
                Text("Open \(Endpoints.remoteHost) on your computer, then approve its pairing code here.")
                    .font(FS.font.bodySm())
                    .foregroundColor(c.textMuted)
                    .fixedSize(horizontal: false, vertical: true)
                pairingCard(c: c)
                approvalCard(c: c)
                activeList(c: c)
                Spacer().frame(height: FS.space.s12)
            }
            .padding(.horizontal, FS.space.s6)
            .padding(.top, FS.space.s4)
        }
        .background(c.bg.ignoresSafeArea())
        .navigationTitle("Remote")
        .navigationBarTitleDisplayMode(.inline)
        .refreshable { await vm.load() }
        .task {
            if case .idle = vm.state { await vm.load() }
            if let initialApprovalLink {
                pastedLink = initialApprovalLink
                _ = vm.stageApproval(link: initialApprovalLink)
            }
        }
        .task {
            while !Task.isCancelled {
                nowMs = Int64(Date().timeIntervalSince1970 * 1000)
                try? await Task.sleep(nanoseconds: 1_000_000_000)
            }
        }
        .sheet(isPresented: $showScanner) {
            CompanionDockScannerSheet(
                onScan: { raw in
                    pastedLink = raw
                    if vm.stageApproval(link: raw) {
                        showScanner = false
                    }
                },
                onError: { scanError = $0 },
                onPasteInstead: { showScanner = false }
            )
        }
        .confirmationDialog(
            pendingRevoke.map { "Disconnect \(displayLabel(for: $0))?" } ?? "Disconnect this browser?",
            isPresented: Binding(
                get: { pendingRevoke != nil },
                set: { if !$0 { pendingRevoke = nil } }
            ),
            titleVisibility: .visible,
            presenting: pendingRevoke
        ) { target in
            Button("Disconnect", role: .destructive) {
                Task {
                    await vm.revoke(tokenPrefix: target.tokenPrefix)
                    pendingRevoke = nil
                }
            }
            Button("Cancel", role: .cancel) { pendingRevoke = nil }
        } message: { _ in
            Text("The browser session will end immediately. It can reconnect any time with a fresh QR.")
        }
    }

    private func pairingCard(c: FSColors) -> some View {
        FSCard {
            VStack(alignment: .leading, spacing: FS.space.s3) {
                Text("Connect a browser")
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundColor(c.text)
                FSPrimaryButton("Scan pairing QR", block: true) {
                    scanError = nil
                    showScanner = true
                }
                .accessibilityIdentifier("companion-dock-scan")
                TextField("Paste pairing link", text: $pastedLink, axis: .vertical)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                    .lineLimit(2...4)
                    .textFieldStyle(.roundedBorder)
                    .accessibilityIdentifier("companion-dock-link")
                FSSecondaryButton("Use pasted link", block: true) {
                    _ = vm.stageApproval(link: pastedLink)
                }
                .disabled(pastedLink.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
                if let scanError {
                    Text(scanError)
                        .font(FS.font.caption())
                        .foregroundColor(c.danger)
                }
            }
        }
    }

    @ViewBuilder
    private func approvalCard(c: FSColors) -> some View {
        if let approval = vm.stagedApproval {
            FSCard {
                VStack(alignment: .leading, spacing: FS.space.s3) {
                    Text("Approve this browser?")
                        .font(.system(size: 15, weight: .semibold))
                        .foregroundColor(c.text)
                    Text(approval.serverDomain)
                        .font(FS.font.mono())
                        .foregroundColor(c.textMuted)
                        .textSelection(.enabled)
                    Text("It will receive a keyless remote session for four hours. Protected actions still require approval from this phone.")
                        .font(FS.font.caption())
                        .foregroundColor(c.textMuted)
                    FSPrimaryButton(vm.approvalPending ? "Approving…" : "Approve", enabled: !vm.approvalPending, block: true) {
                        Task { await vm.approve() }
                    }
                    .accessibilityIdentifier("companion-dock-approve")
                    FSGhostButton("Cancel", block: true) { vm.clearApproval() }
                }
            }
        }
        if vm.approvalComplete {
            FSCard {
                HStack(spacing: FS.space.s2) {
                    Image(systemName: "checkmark.circle.fill").foregroundColor(c.success)
                    Text("Browser connected").font(FS.font.bodySm()).foregroundColor(c.text)
                }
            }
        }
        if let error = vm.approvalError {
            Text(error)
                .font(FS.font.caption())
                .foregroundColor(c.danger)
                .accessibilityIdentifier("companion-dock-approval-error")
        }
    }

    @ViewBuilder
    private func activeList(c: FSColors) -> some View {
        VStack(alignment: .leading, spacing: FS.space.s3) {
            Text("CONNECTED BROWSERS")
                .font(.system(size: 12, weight: .semibold))
                .tracking(1)
                .foregroundColor(c.textMuted)
            switch vm.state {
            case .idle, .loading:
                FSCard {
                    HStack { ProgressView(); Spacer() }
                }
            case .failed(let msg):
                FSCard {
                    VStack(alignment: .leading, spacing: FS.space.s2) {
                        Text("Couldn't load connected browsers").foregroundColor(c.danger)
                        Text(msg).font(FS.font.caption()).foregroundColor(c.textMuted)
                    }
                }
            case .loaded(let list):
                if list.companions.isEmpty {
                    FSCard {
                        HStack(alignment: .top, spacing: FS.space.s2) {
                            Image(systemName: "laptopcomputer").foregroundColor(c.textMuted)
                            VStack(alignment: .leading, spacing: 4) {
                                Text("No browsers connected").font(FS.font.bodySm()).foregroundColor(c.text)
                                Text("Open \(Endpoints.remoteHost) on a computer to add one.")
                                    .font(FS.font.caption())
                                    .foregroundColor(c.textMuted)
                            }
                        }
                    }
                } else {
                    VStack(spacing: FS.space.s3) {
                        ForEach(list.companions) { companion in
                            companionRow(companion, c: c)
                        }
                    }
                }
            }
        }
    }

    private func companionRow(_ companion: CompanionSummary, c: FSColors) -> some View {
        let pending = vm.revokePending.contains(companion.tokenPrefix)
        return FSCard {
            HStack(alignment: .top, spacing: FS.space.s3) {
                Image(systemName: "laptopcomputer")
                    .foregroundColor(c.primary)
                    .imageScale(.large)
                VStack(alignment: .leading, spacing: 2) {
                    Text(displayLabel(for: companion))
                        .font(.system(size: 15, weight: .semibold))
                        .foregroundColor(c.text)
                    Text("last seen \(relative(ms: companion.lastSeenMs))")
                        .font(FS.font.caption())
                        .foregroundColor(c.textMuted)
                    Text(expiresLabel(for: companion))
                        .font(FS.font.caption())
                        .foregroundColor(c.textMuted)
                        .accessibilityIdentifier("companion-expires-\(companion.tokenPrefix)")
                }
                Spacer()
                Button {
                    pendingRevoke = companion
                } label: {
                    if pending {
                        ProgressView()
                    } else {
                        Text("Revoke")
                            .font(.system(size: 14, weight: .semibold))
                            .foregroundColor(c.danger)
                    }
                }
                .disabled(pending)
                .accessibilityIdentifier("companion-revoke-\(companion.tokenPrefix)")
            }
        }
    }

    private func displayLabel(for c: CompanionSummary) -> String {
        return "Session \(c.tokenPrefix)"
    }

    private func expiresLabel(for c: CompanionSummary) -> String {
        let remaining = c.expiresAt - nowMs
        if remaining <= 0 { return "expired" }
        let mins = remaining / 60_000
        if mins >= 60 {
            let hours = mins / 60
            let leftover = mins % 60
            if leftover == 0 { return "expires in \(hours)h" }
            return "expires in \(hours)h \(leftover)m"
        }
        if mins >= 1 { return "expires in \(mins)m" }
        let secs = max(0, remaining / 1000)
        return "expires in \(secs)s"
    }

    private func relative(ms: Int64) -> String {
        Date.flagshipFormatted(epochMs: ms)
    }
}

private struct CompanionDockScannerSheet: View {
    @Environment(\.dismiss) private var dismiss
    let onScan: (String) -> Void
    let onError: (String) -> Void
    let onPasteInstead: () -> Void

    var body: some View {
        NavigationStack {
            VStack(spacing: FS.space.s4) {
                QRScannerView(
                    onScan: onScan,
                    onError: onError,
                    validate: { CompanionDockApprovalLink.parse($0) != nil }
                )
                .clipShape(RoundedRectangle(cornerRadius: FS.radius.md))
                Button("Paste link instead") {
                    onPasteInstead()
                    dismiss()
                }
            }
            .padding(FS.space.s4)
            .navigationTitle("Scan pairing QR")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                }
            }
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/ui/screens/CompanionDockScreen.kt
// P14 — Settings → Dock a browser.
//
// Phone-side companion-dock surface. The desktop starts the ceremony and
// displays a QR; this keyholder phone scans/pastes it, biometric-gates the
// approval, and grants a 4-hour keyless companion session. Active
// companions are listed below; each row has a single-tap revoke with a
// confirm dialog.
//
// Mirrors FlagshipUI/Screens/CompanionDockScreen.swift + the canonical
// webapp `views/companion-dock.js`.

package com.flagshipserver.app.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.flagshipserver.app.api.CompanionSummary
import com.flagshipserver.app.core.LocalScreensClient
import com.flagshipserver.app.core.LocalAppState
import com.flagshipserver.app.core.LocalToastCenter
import com.flagshipserver.app.ui.components.FSCard
import com.flagshipserver.app.ui.components.FSGhostButton
import com.flagshipserver.app.ui.components.FSPrimaryButton
import com.flagshipserver.app.ui.theme.FS
import com.flagshipserver.app.viewmodels.CompanionDockViewModel
import com.flagshipserver.app.viewmodels.LoadingState
import kotlinx.coroutines.delay
import com.flagshipserver.app.core.FlagshipDateFormat
import java.util.Locale

@Composable
fun CompanionDockScreen(nav: NavController, initialLink: String = "") {
    val client = LocalScreensClient.current
    val app = LocalAppState.current
    val toasts = LocalToastCenter.current
    val vm = remember(client, app.currentPod?.fqdn) {
        CompanionDockViewModel(client, expectedServerDomain = app.currentPod?.fqdn)
    }
    val state by vm.state.collectAsState()
    val stagedApproval by vm.stagedApproval.collectAsState()
    val approvalPending by vm.approvalPending.collectAsState()
    val approvalError by vm.approvalError.collectAsState()
    val approvalComplete by vm.approvalComplete.collectAsState()
    val revokePending by vm.revokePending.collectAsState()
    var pendingRevoke by remember { mutableStateOf<CompanionSummary?>(null) }
    var pastedLink by remember { mutableStateOf(initialLink) }
    var scanning by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        vm.load()
        if (initialLink.isNotBlank()) vm.stageApproval(initialLink)
    }

    val scroll = rememberScrollState()
    Column(
        Modifier
            .fillMaxSize()
            .verticalScroll(scroll)
            .padding(horizontal = FS.space.s6),
    ) {
        Spacer(Modifier.height(FS.space.s8))
        FSGhostButton(label = "← Back", onClick = { nav.popBackStack() })
        Spacer(Modifier.height(FS.space.s3))
        Text(
            "Dock a browser",
            color = FS.colors.text,
            style = TextStyle(fontSize = 28.sp, lineHeight = 36.sp, fontWeight = FontWeight.Medium),
        )
        Text(
            "Open web.flagshipserver.com/dock on your computer, then approve its pairing code here.",
            color = FS.colors.textMuted,
            style = TextStyle(fontSize = 13.sp),
        )

        Spacer(Modifier.height(FS.space.s4))

        FSCard(padding = PaddingValues(FS.space.s4)) {
            Column(verticalArrangement = Arrangement.spacedBy(FS.space.s3)) {
                FSPrimaryButton(
                    label = if (scanning) "Close scanner" else "Scan pairing QR",
                    onClick = { scanning = !scanning },
                    block = true,
                )
                if (scanning) {
                    QRScanner(onScanned = { raw ->
                        if (vm.stageApproval(raw)) {
                            pastedLink = raw
                            scanning = false
                        }
                    })
                }
                OutlinedTextField(
                    value = pastedLink,
                    onValueChange = { pastedLink = it },
                    label = { Text("Paste pairing link") },
                    modifier = Modifier.fillMaxWidth(),
                    minLines = 2,
                    maxLines = 4,
                )
                FSGhostButton(
                    label = "Use pasted link",
                    onClick = { vm.stageApproval(pastedLink) },
                    enabled = pastedLink.isNotBlank(),
                    block = true,
                )
            }
        }

        stagedApproval?.let { approval ->
            Spacer(Modifier.height(FS.space.s3))
            FSCard(padding = PaddingValues(FS.space.s4)) {
                Column(verticalArrangement = Arrangement.spacedBy(FS.space.s2)) {
                    Text("Approve this browser?", color = FS.colors.text, fontWeight = FontWeight.SemiBold)
                    Text(approval.serverDomain, color = FS.colors.textMuted, fontFamily = FontFamily.Monospace)
                    Text(
                        "It will receive a keyless companion session for four hours. Protected actions still require approval from this phone.",
                        color = FS.colors.textMuted,
                        style = TextStyle(fontSize = 13.sp),
                    )
                    FSPrimaryButton(
                        label = if (approvalPending) "Approving…" else "Approve with biometrics",
                        onClick = { vm.approve() },
                        enabled = !approvalPending,
                        block = true,
                    )
                    FSGhostButton(label = "Cancel", onClick = { vm.clearApproval() }, block = true)
                }
            }
        }
        if (approvalComplete) {
            Spacer(Modifier.height(FS.space.s3))
            FSCard(padding = PaddingValues(FS.space.s4)) {
                Text("Browser docked", color = FS.colors.text, fontWeight = FontWeight.SemiBold)
            }
        }
        approvalError?.let {
            Spacer(Modifier.height(FS.space.s2))
            Text(it, color = FS.colors.danger, style = TextStyle(fontSize = 13.sp))
        }

        Spacer(Modifier.height(FS.space.s6))
        Text(
            "ACTIVE BROWSERS",
            color = FS.colors.textMuted,
            style = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.SemiBold, letterSpacing = 1.sp),
        )
        Spacer(Modifier.height(FS.space.s2))

        when (val s = state) {
            is LoadingState.Loaded -> {
                if (s.value.companions.isEmpty()) {
                    EmptyCompanionsCard()
                } else {
                    s.value.companions.forEach { c ->
                        CompanionRow(
                            companion = c,
                            revoking = c.tokenPrefix in revokePending,
                            onRevokeRequest = { pendingRevoke = c },
                        )
                        Spacer(Modifier.height(FS.space.s2))
                    }
                }
            }
            is LoadingState.Failed -> ErrorCard(message = s.message, onRetry = { vm.load() })
            else -> ServerCardSkeleton()
        }

        Spacer(Modifier.height(FS.space.s12))
    }

    pendingRevoke?.let { c ->
        AlertDialog(
            onDismissRequest = { pendingRevoke = null },
            title = { Text("Revoke this browser?") },
            text = {
                Text(
                    "Revoke session ${c.tokenPrefix}. The browser session will stop immediately. " +
                        "You can pair it again any time with a fresh QR from the desktop dock page.",
                )
            },
            confirmButton = {
                TextButton(onClick = {
                    val target = c.tokenPrefix
                    pendingRevoke = null
                    vm.revoke(target)
                    toasts.success("Browser revoked.")
                }) {
                    Text("Revoke", color = FS.colors.danger)
                }
            },
            dismissButton = {
                TextButton(onClick = { pendingRevoke = null }) { Text("Cancel") }
            },
        )
    }
}

@Composable
private fun CompanionRow(
    companion: CompanionSummary,
    revoking: Boolean,
    onRevokeRequest: () -> Unit,
) {
    val now by produceState(initialValue = System.currentTimeMillis(), companion.expiresAt) {
        while (true) {
            value = System.currentTimeMillis()
            delay(1_000)
        }
    }
    val remaining = (companion.expiresAt - now).coerceAtLeast(0L)
    FSCard(padding = PaddingValues(FS.space.s4)) {
        Row(verticalAlignment = Alignment.Top) {
            Column(Modifier.weight(1f)) {
                Text(
                    "Session ${companion.tokenPrefix}",
                    color = FS.colors.text,
                    style = TextStyle(fontSize = 15.sp, fontWeight = FontWeight.SemiBold),
                )
                Text(
                    "last seen ${fmtRelative(companion.lastSeenMs, now)} · expires in ${formatRemaining(remaining)}",
                    color = FS.colors.textMuted,
                    style = TextStyle(fontSize = 13.sp),
                )
                Text(
                    "token ${companion.tokenPrefix}…",
                    color = FS.colors.textMuted,
                    style = TextStyle(fontSize = 12.sp, fontFamily = FontFamily.Monospace),
                )
                companion.userAgent?.let {
                    Text(
                        it,
                        color = FS.colors.textMuted,
                        style = TextStyle(fontSize = 12.sp),
                    )
                }
            }
            FSGhostButton(
                label = if (revoking) "Revoking…" else "Revoke",
                onClick = onRevokeRequest,
                enabled = !revoking,
            )
        }
    }
}

@Composable
private fun EmptyCompanionsCard() {
    FSCard(padding = PaddingValues(FS.space.s4)) {
        Text(
            "No browsers docked yet. Open the dock page on a computer to add one.",
            color = FS.colors.textMuted,
            style = TextStyle(fontSize = 13.sp),
        )
    }
}

private fun msUntil(expiresAt: Long): Long =
    (expiresAt - System.currentTimeMillis()).coerceAtLeast(0L)

private fun formatRemaining(ms: Long): String {
    if (ms <= 0L) return "expired"
    val totalSec = ms / 1000
    val hours = totalSec / 3600
    val minutes = (totalSec % 3600) / 60
    val seconds = totalSec % 60
    return when {
        hours > 0 -> String.format(Locale.US, "%dh %02dm", hours, minutes)
        minutes > 0 -> String.format(Locale.US, "%dm %02ds", minutes, seconds)
        else -> "${seconds}s"
    }
}

private fun fmtRelative(at: Long, now: Long): String {
    if (at <= 0L) return "—"
    return FlagshipDateFormat.format(at, nowMs = now)
}
